include/operations/blas3_trees.hpp

```diff
@@ -169,6 +170,303 @@ inline bool do_check<false>(bool) {
+  //! @brief A boolean parameter represents wheather or not matrix A is
+  //! transposed
+  static constexpr bool trans_a = TransA;
+  //! @brief A boolean parameter represents wheather or not matrix B is
+  //! transposed
+  static constexpr bool trans_b = TransB;
+  //! @brief The device cacheline size
+  static constexpr IndexType cl_size = ClSize;
+  //! @brief Number of elements which fit within a cache line.
+  static constexpr IndexType cl_elems = cl_size / sizeof(T);
+  //! @brief Number of work items within a work group
+  static constexpr IndexType wg_size = wg_rows * wg_cols;
+
+  static_assert(wg_cols * item_rows == item_cols * wg_rows,
+                "Work group size should be a multiple "
+                "of the number of rows in a block\n"
+                " --- this is ensured iff: item_rows | wg_cols");
+
+  RHS0 _A;
+  RHS0 _B;
+  RHS1 _C;
+  T alpha;
+  T beta;
+  IndexType m;
+  IndexType n;
+  IndexType k;
+  IndexType lda;
+  IndexType ldb;
+  IndexType ldc;
+
+  inline NoLocalGemmFactory(RHS0 A, RHS0 B, RHS1 C, T alpha, T beta)
+      : _A(A),
+        _B(B),
+        _C(C),
+        alpha(alpha),
+        beta(beta),
+        m(_A.getSizeR()),
+        n(_B.getSizeC()),
+        k(_A.getSizeC()),
+        lda(_A.getSizeL()),
+        ldb(_B.getSizeL()),
+        ldc(_C.getSizeL()) {}
+
+  static inline std::string get_type_string() noexcept {
+    return std::string("NoLocalGemmFactory<") + std::to_string(wg_size) + ", " +
+           type_string<value_type>::get_value() + ">";
+  }
+
+  static inline cl::sycl::nd_range<1> get_nd_range(IndexType m,
+                                                   IndexType n) noexcept {
+    const cl::sycl::range<1> nwg(((m - 1) / (item_rows * wg_rows) + 1) *
+                                 ((n - 1) / (item_cols * wg_cols) + 1));
+    const cl::sycl::range<1> wgs(wg_size);
+
+    return cl::sycl::nd_range<1>(nwg * wgs, wgs);
+  }
+
+  inline IndexType getSize() const { return m * n; }
+
+  inline bool valid_thread(cl::sycl::nd_item<1> ndItem) const {
+    return ((ndItem.get_global_id(0) < getSize()));
+  }
+
+  inline void eval(cl::sycl::nd_item<1> id) noexcept {
+    auto A = _A.getData().get_pointer().get();
+    auto B = _B.getData().get_pointer().get();
+    auto C = _C.getData().get_pointer().get();
+    const auto tile_per_row = ((m - 1) / block_rows) + 1;
+
+    const auto wg_id = id.get_group(0);
+    const auto item_id = id.get_local_id(0);
+    const auto tile_id_row = wg_id % tile_per_row;
+    const auto tile_id_col = wg_id / tile_per_row;
+    const auto local_item_id_row = item_id % wg_rows;
+    const auto local_item_id_col = item_id / wg_rows;
+    const auto wg_row = tile_id_row * block_rows;
+    const auto wg_col = tile_id_col * block_cols;
+    value_type reg_res[item_rows][item_cols] = {};
+    value_type reg_a[item_rows];
+    value_type reg_b[item_cols];
+
+    /*! @brief exiting from any threads outside of the m and n boundary */
```

Comment:
@brief is only for starting a function or a method. This is just a normal comment

Reply:
removed